gap/attr.gi

```diff
@@ -290,6 +290,177 @@ function(D)
+    x[S] := infinity;
+    # Get the subgraph induced by the vertex subset.
+    induced_subgraph := InducedSubdigraph(D, s);
+    # Iterate over the maximal independent sets of D[S]
+    for I in DigraphMaximalIndependentSets(induced_subgraph) do
+        # Create a clone of the subset that we can remove the current
+        # independent set from.
+        s_without_I := ShallowCopy(s);
+        SubtractSet(s_without_I, SetX(I,
+        # Need to relable the independent set back to the original labels.
+        x -> DigraphVertexLabel(induced_subgraph, x)));
+        # Index S \ I
+        i := Sum(s_without_I, x -> 2 ^ (x - 1)) + 1;
+        # The chromatic number of this subset is the minimum value of all
+        # the maximal independent subsets of D[S].
+        if x[i] + 1 < x[S] then
+            x[S] := x[i] + 1;
+        fi;
+    od;
+  od;
+  return x[2 ^ n];
+end
+);
+
+BindGlobal("DIGRAPHS_UnderThreeColourable",
+function(D)
+  local nr, I;
+  nr := DigraphNrVertices(D);
+  if DigraphHasLoops(D) then
+    ErrorNoReturn("the argument <D> must be a digraph with no loops,");
+  elif nr = 0 then
+    return 0;  # chromatic number = 0 iff <D> has 0 verts
+  elif IsNullDigraph(D) then
+    return 1;  # chromatic number = 1 iff <D> has >= 1 verts & no edges
+  elif IsBipartiteDigraph(D) then
+    return 2;  # chromatic number = 2 iff <D> has >= 2 verts & is bipartite
+               # <D> has at least 2 vertices at this stage
+  fi;
+  # Now check if the graph is three colourable. This is done by searching
+  # for a maximal independent set where the subgraph induced by removing the
+  # vertices in the set is bipartite.
+  # Need to make a copy in case we are given a mutable digraph
+  D := DigraphImmutableCopy(D);
+  for I in DigraphMaximalIndependentSets(D) do
+    # Check if removing these vertices gives you a bipartite digraph
+    if IsBipartiteDigraph(DigraphRemoveVertices(D, I)) then
+      return 3;
+    fi;
+  od;
+  # This graph is greater than 3 colourable.
+  return infinity;
+end
+);
+
+InstallMethod(ChromaticNumber, "for a digraph and colouring algorithm",
+[IsDigraph, IsDigraphColouringAlgorithm and IsDigraphColouringAlgorithmByskov],
+function(D, Byskov)
+  local n, a, vertices, x, s, i, j, I, s_copy, subset_iter, induced_subgraph,
+  index_subsets, vertex_copy;
+
+  n := DigraphNrVertices(D);
+  if DigraphHasLoops(D) then
+    ErrorNoReturn("the argument <D> must be a digraph with no loops,");
+  elif n = 0 then
+    return 0;  # chromatic number = 0 iff <D> has 0 verts
+  elif IsNullDigraph(D) then
+    return 1;  # chromatic number = 1 iff <D> has >= 1 verts & no edges
+  elif IsBipartiteDigraph(D) then
+    return 2;  # chromatic number = 2 iff <D> has >= 2 verts & is bipartite
+               # <D> has at least 2 vertices at this stage
+  fi;
+  vertices := DigraphVertices(D);
+  # Store current best colouring for each subset
+  x := [1 .. 2 ^ n];
+  # Empty set is 0 colourable
+  x[1] := 0;
```

Comment:
Maybe double-check this is the empty set?

Reply:
Done!